src/Generator/EnumGenerator/Cases/CaseFactory.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace Laminas\Code\Generator\EnumGenerator\Cases;
+
+use Exception;
+use Laminas\Code\Generator\EnumGenerator\CasesInterface;
+
+use function array_key_exists;
+
+final class CaseFactory
+{
+    /**
+     * @psalm-param array{
+     *      name: string,
+     *      pureCases?: list<string>,
+     *      backedCases?: array{
+     *          type: string,
+     *          cases: array<string, int|string>
+     *      },
+     *      methods?: list<array{
+     *          name: string,
+     *          visibility?: string,
+     *          static?: bool,
+     *          returntype?: string,
+     *      }>
+     * } $options
+     */
+    public static function fromOptions(array $options): CasesInterface
+    {
+        $isPureCasesSpecified   = array_key_exists('pureCases', $options);
+        $isBackedCasesSpecified = array_key_exists('backedCases', $options);
+
+        if ($isPureCasesSpecified && $isBackedCasesSpecified === false) {
+            return self::fromPureCases($options['pureCases'] ?? []);
+        }
+
+        if ($isPureCasesSpecified === false && $isBackedCasesSpecified) {
+            return self::fromBackedCases($options['backedCases'] ?? ['type' => '', 'cases' => []]);
+        }
+
+        /** @todo add more specific exception */
+        throw new Exception('Either `pureCases` or `backedCases` must be specified.');
```

Comment:
We can avoid this scenario at type level.

The docblock would need to be exploded a bit:

```php
    /**
     * @psalm-param array{
     *      name: string,
     *      pureCases: list<string>,
     *      methods?: list<array{
     *          name: string,
     *          visibility?: string,
     *          static?: bool,
     *          returntype?: string,
     *      }>
     * }|array{
     *      name: string,
     *      backedCases: array{
     *          type: string,
     *          cases: array<string, int|string>
     *      },
     *      methods?: list<array{
     *          name: string,
     *          visibility?: string,
     *          static?: bool,
     *          returntype?: string,
     *      }>
     * } $options
     */
```

Reply:
But API consumers could still specify both properties in their implementations when using `EnumGenerator::fromConfig()`, wouldn't it be safer to throw the exception if that happens?

Even after implementing this dockblock, the analysis complains that the offsets of `pureCases` and `backedCases` could be undefined in their respective lines, even though we are checking beforehand.

Reply:
> could be undefined in their respective lines, even though we are checking beforehand.

Yes, that is because type inference can't understand if the left or right trees (in the suggested type above) are used.

That could be an `assert(array_key_exists(...))` after the first conditional:

```php
if (array_key_exists(...)) {
    return ...;
}

assert(array_key_exists(...));
return ...;
```

Reply:
indeed, moving them inline worked.